### DISCUSSION TITLE
Stream on scalar lists

### DISCUSSION THREAD

**Discussion by robrichard** [graphql/defer-stream-wg#21] [Architectural Decision Records]
## Context
Originally posted by @michaelstaib https://github.com/graphql/graphql-spec/pull/742#pullrequestreview-753768653

> When streaming a field that has a list of composite types the iterator will yield objects like the following:

```
{
  "id": "1000",
   "name": "Luke Skywalker"
}
```
>
> These will than be integrated into a result that could look like the following:
>
```
{
  "path": ["hero", "friends", "nodes", 1],
  "data": {
    "id": "1000",
    "name": "Luke Skywalker"
  },
  "hasNext": false
}
```
> This works quite well... But what if the list is a list of leaf values like a list of strings? I have the same issue with a nested list where the element is actually a list of something.
>
> Can data in this case be a scalar or a list? Would this break the contract for GraphQL results?

## Decision 
Current implementation in graphql-js returns scalar fields. It needs to be clarified in the spec PR that this is possible

```
{
  data: {
    scalarList: ['apple'],
  },
  hasNext: true,
},
{
  data: 'banana',
  path: ['scalarList', 1],
  hasNext: true,
},
{
  data: 'coconut',
  path: ['scalarList', 2],
  hasNext: false,
},
```

## Consequences
Clients should not assume `data` will always be a map type when `@stream` is used on scalar fields.

## Status
- [x] defined in spec
- [x] implemented in graphql-js

**Comment by robrichard** [comment]
Discussed during [Dec 2021 WG meeting](https://github.com/graphql/graphql-wg/blob/main/notes/2021-12-02.md#deferstream-updatesissuesdiscussions-30m-rob-richard). Agreement that `data` can be a scalar (or a list for multidimensional list fields). Spec can be updated to define data not as a map, but as the type of the referenced field (or query type for root field).

**Comment by glasser** [comment]
Note for posterity: this has been changed by #32 (and may change more by #42)
